improvement: Add retries on fetch failure

thibmeu opened the pull request:
Wrap events suceptible to network failures around a retry function

Closes #34

armfazh commented on src/index.ts:
```diff
@@ -47,7 +48,7 @@ export const handleTokenRequest = async (ctx: Context, request: Request) => {
 
 	const keyID = tokenRequest.truncatedTokenKeyId.toString();
-	const key = await ctx.cache.ISSUANCE_KEYS.get(keyID);
+	const key = await asyncRetries(ctx, ctx.cache.ISSUANCE_KEYS.get)(keyID);
```
To safely pass methods of a class, one has to bind the class instance before passing the method to the wrapper function
https://javascript.info/bind#solution-2-bind


```suggestion
	const key = await asyncRetries(ctx, ctx.cache.ISSUANCE_KEYS.get.bind(ctx.cache.ISSUANCE_KEYS))(keyID);
```

thibmeu replied:
this was the reason to use arrow function attributes
can do both

armfazh commented on src/utils/promises.ts:
```diff
@@ -0,0 +1,32 @@
+				if (i === retries - 1) {
+					ctx.metrics.asyncRetriesTotal.inc({
+						retries: i,
+						success: false.toString(),
```
```suggestion
						success: 'false',
```

thibmeu replied:
I considered both. Really no preference between the two, except I wanted to avoid ad-hoc strings

armfazh reviewed (commented):
double check the increment of counters by one (instead of by `i`)

thibmeu commented on src/utils/promises.ts:
```diff
@@ -0,0 +1,32 @@
+		for (let i = 0; i < retries; i += 1) {
+			try {
+				const result = await f(...args);
+				ctx.metrics.asyncRetriesTotal.inc({
```
probably actually only recording one retry per call would be great

thibmeu replied:
added a way to path `Labels`, and I'm passing `{ operation: '...' }`

thibmeu commented on src/utils/promises.ts:
```diff
@@ -0,0 +1,32 @@
+				});
+				return result;
+			} catch (e: unknown) {
+				if (i === retries - 1) {
+					ctx.metrics.asyncRetriesTotal.inc({
+						retries: i,
+						success: false.toString(),
+					});
+					throw e;
+				}
```
yes, that's what I'll do to correctly capture metric only once

thibmeu commented:
Updated based on @armfazh comments:
* return outside of the for-loop in `asyncRetries`
* add a label for the operation that's being retried
* change `true`/`false` to string for labels
* bind instead of using arrow methods as attributes

Tests are still passing

thibmeu commented:
updated following @fisherdarling 's ask to wrap `ISSUANCE_KEYS` into an object with cache and retries.
I've added an option `{shouldUseCache: boolean}` to bypass where needed.
